### ISSUE TITLE
[feature request] ability for plugins to modify the pages instances and page types after they have been initialized

### ISSUE THREAD

**Issue by Moonlight63** [getgrav/grav#2055]
I'd like to make a plugin that changes what pages and page types are available based on the user that is logged in. Currently I do this in the admin panel by overriding the twig template, but that is a really bad solution as it blocks other plugins from doing the same. On top of that, it's really troublesome to do on a page-by-page, group-by-group basis site wide for finer control. It really needs to be done programatically with php events. I also tried making a plugin that simply allows me to extend the base Pages.php class, only to find out that it was "frozen".... 

My proposed solution is either unfreezing Pages.php so it may be extended and re-inserted into $grav['pages'] value, or moving the onGetPageBlueprints/onGetPageTemplates/onTwigSiteVariables events to fire **after** blueprints and templates have been scanned and registered (and allow the event to modify the resulting arrays) and after the twig-vars (theme, pages, page, header, etc) have been set but before they are passed to twig.  Or possibly the safer route of adding 2 new events to maintain compatibility while still allowing this function for those who would use it.

For example:

In Twig.php
From:
```
public function processSite($format = null, array $vars = [])
    {
        // set the page now its been processed
        $this->grav->fireEvent('onTwigSiteVariables');
        $pages = $this->grav['pages'];
        $page = $this->grav['page'];
        $content = $page->content();

        $twig_vars = $this->twig_vars;

        $twig_vars['theme'] = $this->grav['config']->get('theme');
        $twig_vars['pages'] = $pages->root();
        $twig_vars['page'] = $page;
        $twig_vars['header'] = $page->header();
        $twig_vars['media'] = $page->media();
        $twig_vars['content'] = $content;
```
To:
```
public function processSite($format = null, array $vars = [])
    {
        // set the page now its been processed
        $pages = $this->grav['pages'];
        $page = $this->grav['page'];
        $content = $page->content();

        $twig_vars = $this->twig_vars;

        $twig_vars['theme'] = $this->grav['config']->get('theme');
        $twig_vars['pages'] = $pages->root();
        $twig_vars['page'] = $page;
        $twig_vars['header'] = $page->header();
        $twig_vars['media'] = $page->media();
        $twig_vars['content'] = $content;

        $event = new Event();
        $event->twig_vars = $twig_vars;
        $this->grav->fireEvent('onLateTwigSiteVariables', $event);
        $twig_vars = $event->twig_vars;
```

And Pages.php from:

```
$scanBlueprintsAndTemplates = function () use ($grav) {
                // Scan blueprints
                $event = new Event();
                $event->types = self::$types;
                $grav->fireEvent('onGetPageBlueprints', $event);

                self::$types->scanBlueprints('theme://blueprints/');

                // Scan templates
                $event = new Event();
                $event->types = self::$types;
                $grav->fireEvent('onGetPageTemplates', $event);

                self::$types->scanTemplates('theme://templates/');
            };
```
To:

```
$scanBlueprintsAndTemplates = function () use ($grav) {
                // Scan blueprints
                self::$types->scanBlueprints('theme://blueprints/');

                $event = new Event();
                $event->types = self::$types;
                $grav->fireEvent('onGetPageBlueprints', $event);
                self::$types = $event->types;

                // Scan templates
                self::$types->scanTemplates('theme://templates/');

                $event = new Event();
                $event->types = self::$types;
                $grav->fireEvent('onGetPageTemplates', $event);
                self::$types = $event->types;
            };
```

The only issue I see with this is a potential security risk from some sort of "rouge plugin" that changes something it shouldn't. The risk from moving the event triggers is pretty small however, as that would only have the ability to change thing temporarily, Allowing modification of the Pages class opens up WAY more power, but also more danger to someone who doesn't know what they are doing.

I can easily submit a pull request with the proposed changes, but I wanted to hear thoughts first.

**Comment by Moonlight63** [comment]
Yes, I understand the efficiency side of it. I am not trying to stop the pages from being processed and cached. All I am after is changing the list of pages that is sent to twig in the "pages" twig var. No re-processing required. 
As an example: In the admin plugin, the pages view shows a list of pages gathered from the "pages" twig var set by Twig.php to $this->grav['pages']->root();. There is currently no way to change that list of pages without changing the pages.html.twig file itself, which I'd like to avoid because multiple plugins modify this file, and whichever one loads last overwrites the previous plugins capabilities. 
Another benefit of allowing these vars to be changed is you only have to change them once, as opposed to modifying every single twig file to use a different var, which would get cumbersome. In particular, if the instances property could be altered after all pages content has been cached but before twig requests any information, like root or parentsRawRoutes, it wouldn't affect the cache at all.

**Comment by Moonlight63** [comment]
I didn't mean to close that....

**Comment by rhukster** [comment]
There will certainly be a performance hit doing things this way.  Let me think on it though.

**Comment by rhukster** [comment]
You have any thoughts on this @mahagr ?

**Comment by mahagr** [comment]
I was trying to avoid participating in this issue as I disagree not only with the proposal, but also the original code. :)

Few notes, though:
1) Renaming or even moving `onTwigSiteVariables` after getting `$page->content()` breaks most of the existing plugins as the function renders Twig by itself and needs those altered variables.
2) If you want to remove page types by a user, you're out of luck as the results get cached. A better way to do this would be to remove the available types in admin, not from the system and risk breaking a bunch of pages in the process.
3) I think you're trying to do the right thing in the wrong place, though there are limitations on how Grav processes things, too.

Not only there are performance considerations on this change, but it also breaks things.

As a side note: I've been experimenting a new way to handle pages, which is both more powerful and faster. Unfortunately, it will still take a while to finish it, so it won't help you for some time.

Is the purpose of this issue to only change administration or also the pages in the site?

**Comment by Moonlight63** [comment]
Yea, I've been playing with things locally trying to figure out how to make it work, and the last thing I wan't is for things to break. It seems if it's going to be handled by events, then new events need to be added rather than changing existing ones. I've had very little success on my end. As you said, things break if not done in the correct order.

The purpose is to change administration. Front end pages are not a problem to limit access to since a plugin can intercept a page request and detect if a user is logged in, and if not, redirect. The reason for not changing the admin plugin directly is simply because it generates all of it's information (at least for pages) right here. I hear what you're saying about right thing wrong place. 

I just want to make it clear that I am not trying to "remove pages and page types" by user. I'm trying to simply limit what a user sees in the admin panel. My current plugin [Users Page Ownership](https://github.com/Moonlight63/grav-plugin-users-page-ownership) does this, but it does it very inefficiently, because every reference to pages, page types, page children in every twig file has to be accounted for and changed. Not to mention other plugins that also add references like this. Every time there is an update to a twig file in the admin panel, it has to be copied and changed in the plugin because most of the visibility features can only be calculated in the twig files themselves. Keeping track of these changes can quickly become overwhelming. 

To give a simple example of the usefulness and need for something like this. Lets say you are running a blog or publication site. You have your site administration, with full access to all site pages, configuration, plugins, etc. You also have multiple writers that need to be able to create posts under a specific "blog" page. Every writer (aka, users in the "authors" group)  needs to be able to see the "blog" page in the pages list view, and assign it as the parent to a post they create, but they can not view it, edit it, move it, or delete it. Under that "blogs" page, they may see links to all of the posts that they have created (and pages that have specifically been shared with them), and they may click these links to view and edit the contents of that page. In the pages view they may or may not see page entries for the other posts under the blogs page, but any page that is not created by them should not be a link, therefore inaccessible to modification. In a perfect world, these pages would be unable to be deleted after they have been published, except by administration.

I don't expect all of this functionality to be available within grav core. Not everyone needs it. The issue is that as the system is now, I truly think it's impossible to come up with an elegant solution that extends the admin plugin directly without hindering other, equally if not more useful plugins, for people like me who spend a lot of time in the admin panel managing lots of pages. I'm doing my best to fix that problem, and if extending twig files directly was more flexible, then it could probably be done. For example, my plugin is completely incompatible with the [core-services-manager](https://github.com/Flamenco/grav-plugin-core-service-manager) plugin, because it also completely replaces most if not all the admin twig files. 
My issue is small, sure, because it would be trivial for me to simply "add support" for core-service-manager and even use it, build on top of it, and require it in my own plugins. But that isn't the point. What happens when another plugin runs into this same problem. There isn't any standardization for what unofficial plugins should modify what files in what order, and plugins really have no way to work both together and separately if they both must modify the same twig file. And to be perfectly clear, that makes perfect sense! Twig is for displaying thing! Not calculation. So why then, can I not tell twig what pages it should have a list of... Why must that calculation be done in twig? [Calc](https://github.com/Moonlight63/grav-plugin-users-page-ownership/blob/master/templates/pages.html.twig#L108-L205) [Render](https://github.com/Moonlight63/grav-plugin-users-page-ownership/blob/master/templates/pages.html.twig#L241-L291)

**TL;DR** I'm by no means trying to rant or write a book. Like I said, my own personal issue is small, and can easily be resolved (even if there might be a better way). I hope that you can understand and appreciate that I am not bringing things like this up to be a burden, but because I am trying to find ways to make things better, more powerful, more flexible. Yes, I am effectively trying to fix a problem that hasn't happened yet, but isn't that a good thing? Even changes that aren't necessary now could become invaluable in the future. I choose to use grav not only because of it's power and flexibility, but also because of it's growing community that continues to only increase it's affability more and more. Truth be told, I've had to think long and hard about this response and what I am trying to accomplish. It is not simply a means to an end. I try not to make requests lightly. My only goal here is to contribute and help improve the extensibility as a whole.

Ok, that TLDR was still long so I think it only appropriate to thank you for bearing with me. I'm very interested to hear more thoughts on the matter if you have the time.

**Comment by mahagr** [comment]
I agree with you 101%. The reason why I asked you what you are trying to do is that I did feel that you were trying to do the right thing in the wrong place. Let me explain why...

The place where you were making the change is global and affects all the users equally as everything which happens in the pages is shared among all users. It's a global state of the site. Right now it kind of works because of caching in admin is isolated from the site, though that is something which may change in the future as it would be really nice to be able to warm up cache right when you save your change instead of forcing full cache clear in your site on every change you make.

But even if the admin is isolated right now, page cache is still being shared between all users using admin, so making changes in there affect all users (you missed the caching of the result). So even now your approach could have side effects on other admin users using the site at the same time. This would be much worse if one day we decided to share the cache with the frontend.

Now you probably understand why I was talking about removing pages/types from the site. I think that we need an alternative approach which does not change the site, but only affects the visibility of things per user. And because of we already have ACL in the site, only admin is lacking this feature. Some of it could be in Grav itself, some in admin code and rest of it in twig templates. But I think there has to be a better way to implement this -- in admin specific way.

I'm leaving this question open for now, as right now I do not have time to look at your code to see what you're doing there. What I'm looking for, though, is admin specific changes which abstract the logic out of the template files while providing the hooks needed to do this without drastic template file changes.

**Comment by Moonlight63** [comment]
> admin specific changes which abstract the logic out of the template files while providing the hooks needed to do this without drastic template file changes.

That EXACTLY!

I've had the night to think about this, and I had a sort of revelation, a slight change in philosophy. Tell me if you agree. 
Pages should have permissions similar to OS file permissions. I imagine 4, possibly 5 permissions:
Can view: Allows it to be used as a parent
Can modify,
Can move: only relevant for ordered children,
Can delete,
Possibly can "see" as in: "This is visible in admin pages view but not a link. Automatically true if can view, modify, delete is true, or if user has a page that is a child of this page where any of these are true. Something like that.

Parent pages could set these page permissions for their children, such as can move, and groups can have overrides. The page creators, by default can modify, etc..

Idk, just thought of it in those terms and liked it.

**Comment by mahagr** [comment]
yeah, basically the pages should have owner(s) and group(s). Right now Grav supports only view level actions and even those only in the site by login plugin. Owners are individual users, groups are just Grav groups.

The issue is that ACL tends to get very complicated very fast, especially if there's any level of inheritance. Joomla is a good example on this as it has very powerful ACL (basically what you said), but nobody knows how to use it. It may be better to have something much simpler in the core but allow it to be expanded if needed.

I wouldn't define individual permissions for the current item, just what it would mean to be an owner or to belong into a group. This logic could be in an event (just like login works) and you could then add your own rules.

**Comment by Moonlight63** [comment]
Right, give pages a hook and let plugins use it. That sound perfect. Figuring out how to implement that.... I mean, just adding a user and group to frontmatter is all it takes, but then the whole rendering the page list in admin in twig is the next challenge... I have no ideas for that...
